DR2-1890 Use mutable head and other fixes.

MancunianSam opened the pull request:
There are a few changes here

The first is the change to use the FIFO queue and the mutable head.
There is a description in the README as to how this works.

The second is to filter out any event actions from Preservica for
command_download. When this process downloads the file for the first
time, it will create another two versions because the download is appended to
the event actions and this changes the checksum of both the IO and CO
metadata. This is a bit silly.

The third is that there's another bug which I thought I'd fixed.
We're deduplicating messages based on the id which then selects a random
message where there are duplicates. We then delete that message but the
message/receipt handle that was removed by `distinctBy` stays in the
queue until it is processed without any duplicates.

The tests won't pass until https://github.com/nationalarchives/da-aws-clients/pull/287 is merged and the dependency version is updated but otherwise this is ready.

techncl commented on custodial-copy-backend/src/main/scala/uk/gov/nationalarchives/custodialcopy/Processor.scala:
```diff
@@ -313,8 +316,9 @@ class Processor(
-    } yield ()
+    } yield messageResponse.message.ref
+
+  def commit(id: UUID): IO[Unit] = ocflService.commit(id)
```
Maybe commitStagedChanges?

techncl reviewed (commented):
Looks good! Just 2 comments plus:

1. should we add an `OcflServiceTest` test for the `commit` method in the OcflService?
2. should we add a `ProcessorTest` test for the `commit` method in the Processor?
3. Is there a way to confirm that `commit` is being called in `main` or is the fact that the `objectVersion` is 2 proof enough?
4. the `process` method of the `Processor` returns `messageResponse.message.ref` now, should we check this in the tests to see if it's returning the correct ref?

techncl commented on custodial-copy-backend/src/main/scala/uk/gov/nationalarchives/custodialcopy/Main.scala:
```diff
@@ -75,36 +76,59 @@ object Main extends IOApp {
+      allFibers = fibersForNonDeletedEntities ++ fibersForDeletedEntities
+      allFiberResults = ndeFiberResults ++ deFiberResults
+
+      _ <- processor.commit(UUID.fromString(groupId))
```
Does this mean that on the 2nd go around, the staged changes will be committed?

MancunianSam commented:
1. Yes we should, I've added two, one if there are staged changes and one if not
2. We should add one here to, I've put one in.
3. It's enough that the version is 2, it wouldn't be if we weren't calling commit.
4. I've added in another test to check it's returning the id that's passed to it.

techncl reviewed (approved):
👍
